file.cc

```diff
@@ -51,182 +33,94 @@ bool tryPossibleZipName( std::string const & name, std::string & copyTo )
-  return stat( filename, &buf ) == 0 || errno == EOVERFLOW;
-#endif
+  return QFileInfo( QString::fromStdString( filename ) ).exists();
 }
 
-void Class::open( char const * filename, char const * mode ) 
+void Class::open( char const * mode )
 {
   QFile::OpenMode openMode = QIODevice::Text;
-  const char * pch = mode;
-  while( *pch )
-  {
-    switch( *pch )
-    {
-      case 'r': openMode |= QIODevice::ReadOnly;
-                break;
-      case 'w': openMode |= QIODevice::WriteOnly;
-                break;
-      case '+': openMode &= ~( QIODevice::ReadOnly | QIODevice::WriteOnly );
-                openMode |= QIODevice::ReadWrite;
-                break;
-      case 'a': openMode |= QIODevice::Append;
-                break;
-      case 'b': openMode &= ~QIODevice::Text;
-                break;
-      default:  break;
+
+  for ( const char * i = mode; *i != '\0'; ++i ) {
+    switch ( *i ) {
+      case 'r':
+        openMode |= QIODevice::ReadOnly;
+        break;
+      case 'w':
+        openMode |= QIODevice::WriteOnly;
+        break;
+      case '+':
+        openMode &= ~( QIODevice::ReadOnly | QIODevice::WriteOnly );
+        openMode |= QIODevice::ReadWrite;
+        break;
+      case 'a':
+        openMode |= QIODevice::Append;
+        break;
+      case 'b':
+        openMode &= ~QIODevice::Text;
+        break;
+      default:
+        break;
     }
-    ++pch;
   }
 
-  f.setFileName( filename );
 
   if ( !f.open( openMode ) )
-    throw exCantOpen( std::string( filename ) + ": " + f.errorString().toUtf8().data() );
+    throw exCantOpen( f.fileName().toStdString() + ": " + f.errorString().toUtf8().data() );
 }
 
-Class::Class( char const * filename, char const * mode ) :
-  writeBuffer( 0 )
+Class::Class( std::string_view filename, char const * mode )
 {
-  open( filename, mode );
+  f.setFileName( QString::fromStdString( filename.data() ) );
+  open( mode );
 }
 
-Class::Class( std::string const & filename, char const * mode )
-  : writeBuffer( 0 )
+void Class::read( void * buf, qint64 size )
 {
-  open( filename.c_str(), mode );
-}
-
-void Class::read( void * buf, qint64 size ) 
-{
-  if ( !size )
-    return;
-
-  if ( writeBuffer )
-    flushWriteBuffer();
-
-  qint64 result = f.read( reinterpret_cast<char *>( buf ), size );
+  qint64 result = f.read( static_cast< char * >( buf ), size );
 
   if ( result != size )
     throw exReadError();
 }
 
-size_t Class::readRecords( void * buf, qint64 size, size_t count ) 
+size_t Class::readRecords( void * buf, qint64 size, qint64 count )
 {
-  if ( writeBuffer )
-    flushWriteBuffer();
-
-  qint64 result = f.read( reinterpret_cast<char *>( buf ), size * count );
+  qint64 result = f.read( static_cast< char * >( buf ), size * count );
   return result < 0 ? result : result / size;
 }
 
-void Class::write( void const * buf, qint64 size ) 
+void Class::write( void const * buf, qint64 size )
 {
-  if ( !size )
-    return;
```

Comment:
I changed it to `assert(size!=0)` rather than silently return.

If this condition ever true, something bad happens.

Reply:
will assert  be ignored in release mode?

Reply:
No, it won't :sweat_smile: 

https://github.com/xiaoyifang/goldendict/blob/54a4a052a5a09dbb6994b4eb58ef2196545be450/goldendict.pro#L30-L32

I think we should remove this line, so that all assert/Q_ASSERT works. This will narrow down crashes.

Reply:
> No, it won't due to the existence of `NO_DEBUG`😅
> 
> https://github.com/xiaoyifang/goldendict/blob/54a4a052a5a09dbb6994b4eb58ef2196545be450/goldendict.pro#L30-L32
> 
> I think we should remove this line, so that all assert/Q_ASSERT works. This will narrow down crashes.
> 
> The `Q_ASSERT` is still useable because we don't define `QT_NO_DEBUG`

This can wait

Reply:
TBH, this size check is not doing anything. If a size=0 got passed here, something catastrophic has happened.

I added exception throwing here. We may consider `assert` later

Reply:
> If a size=0 got passed here, something catastrophic has happened.

I think size=0 will not got passed ,it got returned(ignored).  I think return directly should be ok.